### PR TITLE
Move logic to makes queries without limits to the relevant adapters

### PR DESCRIPTION
**Pull request by NMe84**
This PR addresses the issue in #236. The problem there was that the `DataTableExporterManager` was hardcoded to set the length of the result set to -1 to indicate an unrestricted data set. This worked for MySQL, but other data sources potentially have different ways to pass a request for an unlimited amount of results. MongoDB specifically requires a length of 0 to indicate you want as many records as match the query's filters. -1 would result in 0 or 1 records to be returned, which obviously isn't helpful for an exporter.

I moved the responsibility for the specifics into the specific Adapter classes. The `DataTableState` now expects `null` to be passed as a length in order for the responsible adapter to translate that into a query without a limit. In order to do this I did need to change some function signatures (particularly to allow null to be passed) so this _does_ include breaking API changes and warrants a new 0.x release rather than a patch.

I added deprecation notices to the `DataTableState::setLength` and `DataTableState::setStart` methods to warn users that they're using outdated or plain wrong arguments and I'd suggest turning those into hard errors in an eventual 1.x release.

### CODE REVIEW

File: src/Adapter/ArrayAdapter.php [hunk 1]
```diff
@@ -72,7 +72,7 @@ public function getData(DataTableState $state): ResultSetInterface
 
         $data = iterator_to_array($this->processData($state, $this->data, $map));
 
-        $length = $state->getLength();
+        $length = $state->getLength() ?? 0;
         $page = $length > 0 ? array_slice($data, $state->getStart(), $state->getLength()) : $data;
```
Comment (curry684) [inline src/Adapter/ArrayAdapter.php:76]: You're ignoring the start value now for unlimited length, seems wrong?

I would expect something like:
```
$page = $array_slice($data, $state->getStart(), $state->getLength() ?? count($data));
```
Which would translate to 'respect $start but max out at length or unlimited if null'.

**Review by curry684 — CHANGES_REQUESTED** [review]
ArrayAdapter is now functioning differently than before.

File: src/Adapter/ArrayAdapter.php (hunk 1 above)
Comment (NMe84) [reply src/Adapter/ArrayAdapter.php:76]: All the other adapters also ignore start if length is not a positive integer, even before I touched them. I assumed this was by design.

Comment (curry684) [reply src/Adapter/ArrayAdapter.php:76]: Hmmm no it seems to be quite inconsistent actually - Mongo does respect the start value currently, ORM doesn't.

I think we should make it consistent one way or another if we're breaking this anyway. Given that I don't see a real use case for respecting `start` with unlimited length I'm fine with ignoring it. Did you accidentally 'fix' this in Mongo?

Comment (NMe84) [reply src/Adapter/ArrayAdapter.php:76]: No, I didn't touch whatever was done with the `start` property other than the sanitation + warning I added to its setter method.

Comment (curry684) [reply src/Adapter/ArrayAdapter.php:76]: Modified Mongo to be consistent with the other adapters. If we break we should break consistently.

**Pull request merged**